From 84264bbc126f5df11a3dab820c43238a43e24756 Mon Sep 17 00:00:00 2001
From: Peter Cai <peter@typeblog.net>
Date: Fri, 24 Apr 2026 08:54:28 -0300
Subject: [PATCH] audio_hal_interface: Optionally use sysbta HAL

Required to support sysbta, our system-side bt audio implementation.

Ported to android-16.0.0_r4 by bariccattion.
---
 .../aidl/a2dp/client_interface_aidl.cc                |  8 ++++----
 .../aidl/a2dp/client_interface_aidl.h                 |  7 +++++++
 .../audio_hal_interface/aidl/client_interface_aidl.cc |  8 ++++----
 .../audio_hal_interface/aidl/client_interface_aidl.h  |  7 +++++++
 system/audio_hal_interface/hal_version_manager.cc     | 11 +++++++++--
 5 files changed, 31 insertions(+), 10 deletions(-)

diff --git a/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.cc b/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.cc
index 201cae87bb..ec238d5717 100644
--- a/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.cc
+++ b/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.cc
@@ -59,7 +59,7 @@ BluetoothAudioClientInterface::~BluetoothAudioClientInterface() {
 bool BluetoothAudioClientInterface::IsValid() const { return provider_ != nullptr; }
 
 bool BluetoothAudioClientInterface::is_aidl_available() {
-  return AServiceManager_isDeclared(kDefaultAudioProviderFactoryInterface.c_str());
+  return AServiceManager_isDeclared(audioProviderFactoryInterface().c_str());
 }
 
 std::vector<AudioCapabilities> BluetoothAudioClientInterface::GetAudioCapabilities() const {
@@ -73,7 +73,7 @@ std::vector<AudioCapabilities> BluetoothAudioClientInterface::GetAudioCapabiliti
     return capabilities;
   }
   auto provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-          AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+          AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
 
   if (provider_factory == nullptr) {
     log::error("can't get capability from unknown factory");
@@ -98,7 +98,7 @@ BluetoothAudioClientInterface::GetProviderInfo(
 
   if (provider_factory == nullptr) {
     provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-            AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+            AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
   }
 
   if (provider_factory == nullptr) {
@@ -174,7 +174,7 @@ void BluetoothAudioClientInterface::FetchAudioProvider() {
   // re-registered, so we need to re-fetch the service.
   for (int retry_no = 0; retry_no < kFetchAudioProviderRetryNumber; ++retry_no) {
     auto provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-            AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+            AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
 
     if (provider_factory == nullptr) {
       log::error("can't get capability from unknown factory");
diff --git a/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.h b/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.h
index e492fde23a..a24ffefedf 100644
--- a/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.h
+++ b/system/audio_hal_interface/aidl/a2dp/client_interface_aidl.h
@@ -28,6 +28,7 @@
 #include "bta/le_audio/broadcaster/broadcaster_types.h"
 #include "bta/le_audio/le_audio_types.h"
 #include "transport_instance.h"
+#include "osi/include/properties.h"
 
 // Keep after audio_aidl_interfaces.h because of <base/logging.h>
 // conflicting definitions.
@@ -164,6 +165,12 @@ protected:
   //     "android.hardware.bluetooth.audio.IBluetoothAudioProviderFactory/default";
   static inline const std::string kDefaultAudioProviderFactoryInterface =
           std::string() + IBluetoothAudioProviderFactory::descriptor + "/default";
+  static inline const std::string kSystemAudioProviderFactoryInterface =
+          std::string() + IBluetoothAudioProviderFactory::descriptor + "/sysbta";
+  static inline const std::string audioProviderFactoryInterface() {
+      return osi_property_get_bool("persist.bluetooth.system_audio_hal.enabled", false)
+       ? kSystemAudioProviderFactoryInterface : kDefaultAudioProviderFactoryInterface;
+  }
 
 private:
   IBluetoothTransportInstance* transport_;
diff --git a/system/audio_hal_interface/aidl/client_interface_aidl.cc b/system/audio_hal_interface/aidl/client_interface_aidl.cc
index c1ce661ff7..401fd5b531 100644
--- a/system/audio_hal_interface/aidl/client_interface_aidl.cc
+++ b/system/audio_hal_interface/aidl/client_interface_aidl.cc
@@ -73,7 +73,7 @@ BluetoothAudioClientInterface::BluetoothAudioClientInterface(IBluetoothTransport
 bool BluetoothAudioClientInterface::IsValid() const { return provider_ != nullptr; }
 
 bool BluetoothAudioClientInterface::is_aidl_available() {
-  return AServiceManager_isDeclared(kDefaultAudioProviderFactoryInterface.c_str());
+  return AServiceManager_isDeclared(audioProviderFactoryInterface().c_str());
 }
 
 std::vector<AudioCapabilities> BluetoothAudioClientInterface::GetAudioCapabilities() const {
@@ -87,7 +87,7 @@ std::vector<AudioCapabilities> BluetoothAudioClientInterface::GetAudioCapabiliti
     return capabilities;
   }
   auto provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-          AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+          AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
 
   if (provider_factory == nullptr) {
     log::error("can't get capability from unknown factory");
@@ -112,7 +112,7 @@ BluetoothAudioClientInterface::GetProviderInfo(
 
   if (provider_factory == nullptr) {
     provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-            AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+            AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
   }
 
   if (provider_factory == nullptr) {
@@ -144,7 +144,7 @@ void BluetoothAudioClientInterface::FetchAudioProvider() {
   // re-registered, so we need to re-fetch the service.
   for (int retry_no = 0; retry_no < kFetchAudioProviderRetryNumber; ++retry_no) {
     auto provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-            AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+            AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
 
     if (provider_factory == nullptr) {
       log::error("can't get capability from unknown factory");
diff --git a/system/audio_hal_interface/aidl/client_interface_aidl.h b/system/audio_hal_interface/aidl/client_interface_aidl.h
index a3c60384ab..ee5901e32a 100644
--- a/system/audio_hal_interface/aidl/client_interface_aidl.h
+++ b/system/audio_hal_interface/aidl/client_interface_aidl.h
@@ -29,6 +29,7 @@
 #include "bta/le_audio/broadcaster/broadcaster_types.h"
 #include "bta/le_audio/le_audio_types.h"
 #include "transport_instance.h"
+#include "osi/include/properties.h"
 
 namespace bluetooth {
 namespace audio {
@@ -150,6 +151,12 @@ protected:
   //     "android.hardware.bluetooth.audio.IBluetoothAudioProviderFactory/default";
   static inline const std::string kDefaultAudioProviderFactoryInterface =
           std::string() + IBluetoothAudioProviderFactory::descriptor + "/default";
+  static inline const std::string kSystemAudioProviderFactoryInterface =
+          std::string() + IBluetoothAudioProviderFactory::descriptor + "/sysbta";
+  static inline const std::string audioProviderFactoryInterface() {
+      return osi_property_get_bool("persist.bluetooth.system_audio_hal.enabled", false)
+       ? kSystemAudioProviderFactoryInterface : kDefaultAudioProviderFactoryInterface;
+  }
 
 private:
   IBluetoothTransportInstance* transport_;
diff --git a/system/audio_hal_interface/hal_version_manager.cc b/system/audio_hal_interface/hal_version_manager.cc
index 1a92c1aeb2..bb27167ca2 100644
--- a/system/audio_hal_interface/hal_version_manager.cc
+++ b/system/audio_hal_interface/hal_version_manager.cc
@@ -20,6 +20,7 @@
 #include <android/hidl/manager/1.2/IServiceManager.h>
 #include <bluetooth/log.h>
 #include <hidl/ServiceManagement.h>
+#include "osi/include/properties.h"
 
 #include <memory>
 #include <mutex>
@@ -34,6 +35,12 @@ using ::aidl::android::hardware::bluetooth::audio::IBluetoothAudioProviderFactor
 
 static const std::string kDefaultAudioProviderFactoryInterface =
         std::string() + IBluetoothAudioProviderFactory::descriptor + "/default";
+static const std::string kSystemAudioProviderFactoryInterface =
+        std::string() + IBluetoothAudioProviderFactory::descriptor + "/sysbta";
+static inline const std::string audioProviderFactoryInterface() {
+    return osi_property_get_bool("persist.bluetooth.system_audio_hal.enabled", false)
+     ? kSystemAudioProviderFactoryInterface : kDefaultAudioProviderFactoryInterface;
+}
 
 std::string toString(BluetoothAudioHalTransport transport) {
   switch (transport) {
@@ -74,7 +81,7 @@ BluetoothAudioHalVersion GetAidlInterfaceVersion() {
   static auto aidl_version = []() -> BluetoothAudioHalVersion {
     int version = 0;
     auto provider_factory = IBluetoothAudioProviderFactory::fromBinder(::ndk::SpAIBinder(
-            AServiceManager_waitForService(kDefaultAudioProviderFactoryInterface.c_str())));
+            AServiceManager_waitForService(audioProviderFactoryInterface().c_str())));
 
     if (provider_factory == nullptr) {
       log::error("getInterfaceVersion: Can't get aidl version from unknown factory");
@@ -139,7 +146,7 @@ android::sp<IBluetoothAudioProvidersFactory_2_0> HalVersionManager::GetProviders
 
 HalVersionManager::HalVersionManager() {
   hal_transport_ = BluetoothAudioHalTransport::UNKNOWN;
-  if (AServiceManager_checkService(kDefaultAudioProviderFactoryInterface.c_str()) != nullptr) {
+  if (AServiceManager_checkService(audioProviderFactoryInterface().c_str()) != nullptr) {
     hal_version_ = GetAidlInterfaceVersion();
     hal_transport_ = BluetoothAudioHalTransport::AIDL;
     return;
-- 
2.53.0.windows.3

